Multiz Rework: Human Suffering Edition (Contains PLANE CUBE)

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

I've reworked multiz. This was done because our current implementation of multiz flattens planes down into just the openspace plane. This breaks any effects we attach to plane masters (including lighting), but it also totally kills the SIDE_MAP map format, which we NEED for wallening (A major 3/4ths resprite of all wall and wall adjacent things, making them more then one tile high. Without sidemap we would be unable to display things both in from of and behind objects on map. Stupid.)

This required MASSIVE changes. Both to all uses of the plane var for reasons I'll discuss later, and to a ton of different systems that interact with rendering.

I'll do my best to keep this compact, but there's only so much I can do. Sorry brother.

### Core idea

OK: first thing.
`vis_contents` as it works now squishes the planes of everything inside it down into the plane of the vis_loc.
This is bad. But how to do better?

It's trivially easy to make copies of our existing plane masters but offset, and relay them to the bottom of the plane above. Not a problem. The issue is how to get the actual atoms on the map to "land" on them properly.

We could use `FLOAT_PLANE` to offset planes based off how they're being seen, in theory this would allow us to create lens for how objects are viewed.
But that's not a stable thing to do, because properly "landing" a plane on a desired plane master would require taking into account every bit of how it's being seen, would inherently break this effect.

Ok so we need to manually edit planes based off "z layer" (IE: what layer of a z stack are you on).

That's the key conceit of this pr. Implementing the plane cube, and ensuring planes are always offset properly.
Everything else is just gravy.

### About the Plane Cube 

Each plane master (except ones that opt out) is copied down by some constant value equal to the max absolute change between the first and the last plane.
We do this based off the max z stack size detected by SSmapping. This is also where updates come from, and where all our updating logic will live.

As mentioned, plane masters can choose to opt out of being mirrored down. In this case, anything that interacts with them assuming that they'll be offset will instead just get back the valid plane value. This works for render targets too, since I had to work them into the system as well.

Plane masters can also be temporarily hidden from the client's screen. This is done as an attempt at optimization, and applies to anything used in niche cases, or planes only used if there's a z layer below you.

### About Plane Master Groups

BYOND supports having different "maps" on screen at once (IE: groups of items/turfs/etc)
Plane masters cannot cover 2 maps at once, since their location is determined by their screen_loc.
So we need to maintain a mirror of each plane for every map we have open.

This was quite messy, so I've refactored it (and maps too) to be a bit more modular.

Rather then storing a list of plane masters, we store a list of plane master group datums.
Each datum is in charge of the plane masters for its particular map, both creating them, and managing them.

Like I mentioned, I also refactored map views. Adding a new mapview is now as simple as newing a `/atom/movable/screen/map_view`, calling `generate_view` with the appropriate map id, setting things you want to display in its `vis_contents`, and then calling `display_to` on it, passing in the mob to show ourselves to.

Much better then the hardcoded pattern we used to use. So much duplicated code man.

Oh and plane master controllers, that system we have that allows for applying filters to sets of plane masters? I've made it use lookups on plane master groups now, rather then hanging references to all impacted planes. This makes logic easier, and prevents the need to manage references and update the controllers.

![image](https://user-images.githubusercontent.com/58055496/183850002-ca75ea26-f6cd-44a1-80f1-29168cd6a0aa.png)

In addition, I've added a debug ui for plane masters.
It allows you to view all of your own plane masters and short descriptions of what they do, alongside tools for editing them and their relays.

It ALSO supports editing someone elses plane masters, AND it supports (in a very fragile and incomplete manner) viewing literally through someone else's eyes, including their plane masters. This is very useful, because it means you can debug "hey my X is yorked" issues yourself, on live.

In order to accomplish this I have needed to add setters for an ungodly amount of visual impacting vars. Sight flags, eye, see_invis, see_in_dark, etc.

It also comes with an info dump about the ui, and plane masters/relays in general.

Sort of on that note. I've documented everything I know that's niche/useful about our visual effects and rendering system. My hope is this will serve to bring people up to speed on what can be done more quickly, alongside making my sin here less horrible.
See https://github.com/LemonInTheDark/tgstation/blob/multiz-hell/.github/guides/VISUALS.md.

### "Landing" planes

Ok so I've explained the backend, but how do we actually land planes properly?
Most of the time this is really simple. When a plane var is set, we need to provide some spokesperson for the appearance's z level. We can use this to derive their z layer, and thus what offset to use.

This is just a lot of gruntwork, but it's occasionally more complex.
Sometimes we need to cache a list of z layer -> effect, and then use that.
Also a LOT of updating on z move. So much z move shit.

Oh. and in order to make byond darkness work properly, I needed to add SEE_BLACKNESS to all sight flags.
This draws darkness to plane 0, which means I'm able to relay it around and draw it on different z layers as is possible. ~~fun darkness ripple effects incoming someday~~

I also need to update mob overlays on move.
I do this by realiizing their appearances, mutating their plane, and then readding the overlay in the correct order.

The cost of this is currently 3N. I'm convinced this could be improved, but I've not got to it yet.
It can also occasionally cause overlays to corrupt. This is fixed by laying a protective ward of `overlays.Copy` in the sand, but that spell makes the compiler confused, so I'll have to bully lummy about fixing it at some point.

### Behavior changes

We've had to give up on the already broken gateway "see through" effect. Won't work without managing gateway plane masters or something stupid. Not worth it.
So instead we display the other side as a ui element. It's worse, but not that bad.

Because vis_contents no longer flattens planes (most of the time), some uses of it now have interesting behavior.
The main thing that comes to mind is alert popups that display mobs. They can impact the lighting plane.
I don't really care, but it should be fixable, I think, given elbow grease.

Ah and I've cleaned up layers and plane defines to make them a bit easier to read/reason about, at least I think.

## Why It's Good For The Game

Behold, you can see multi multizs 
![image](https://user-images.githubusercontent.com/58055496/183854909-3020d1e8-eb6b-4726-9bc7-3996a303731b.png)
![image](https://user-images.githubusercontent.com/58055496/183854514-2e3360d9-02dd-4fc7-863f-65f61b17c0a2.png)
Also dropshadow and lighting works now
![image](https://user-images.githubusercontent.com/58055496/183854954-ba2ec644-76e2-43e8-9430-b8b65005fc6a.png)
![image](https://user-images.githubusercontent.com/58055496/183854553-a78193ab-11d4-4023-9f7e-0fb0f235bec9.png)

Fixes #65800
Fixes #68461 
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
refactor: Refactored... well a lot really. Map views, anything to do with planes, multiz, a shit ton of rendering stuff. Basically if you see anything off visually report it
admin: VV a mob, and hit View/Edit Planes in the dropdown to steal their view, and modify it as you like. You can do the same to yourself using the Edit/Debug Planes verb
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

LemonInTheDark committed:
Implements the plane cube backend

Ok so like, right now, we have an issue with multiz
Effects applied to planes displayed on lower z levels break when piped
through multiz.

This is because we use a vis_contents, and set everything's vis flags to
PARENT_PLANE.

This essentially tells the turfs and mobs we rerender to lose their
plane, and just render into the plane of the turf they're being copied
into.

This is typically mostly unoticable, but it absolutely borks side_map,
since it uses planes to separate things into uninterlayerable groups,
needs to since it does more advanced sorting based off the thing's
position.

So, the solution.

Instead of having just one set of plane masters, we have one for each
max "layer" of z interlocking in the game.

Rather then piping the lower layer's plane masters into the client's
view, we instead pipe them onto the openspace plane of the layer above.

This lets us apply effects to just X level, and also fixes this effects
failing issue.

I've attempted to implement this in such a way that reduces the impact
of the typical user. I will need to add better getters, and do a pass of
all uses of plane to either use float_plane, or to update on z change.

These should? be doable tho. I have hope

LemonInTheDark committed:
Does a pass on all /atom uses of plane, misc fixes

Converted all non hud (because you never actually want your hud elements
to underlay) of plane to the new system.
Ensures plane is properly set between moves
Adds hooks to track a client's eye plane offset, uses them to fix a bug
with the lighting plane
Adds a macro for getting all the planes in a group, uses it to make all
uses of plane master getting apply to all plane masters in a "type"
Adds support for plane masters rendering towards multiple targets. Plan
on using this in future
Shortens the plane master range, it was causing plane values to dip
below -10000, and brick the renderer
Makes some overlays "follow" their source, should really add a more
generic way of doing this

LemonInTheDark committed:
Some general fixes and cleanup

Fixes misplaned'd area lighting
Fixes new planes potentially being set to 0 if a move occured before
lists were built
Ensures parallax will only ever render on the bottom of a "z stack"
Fixes weird parallax behavior, turns out negative layers are "float"
layers, and were breaking it

LemonInTheDark committed:
Ensures the secure eye program does NOT attempt to load relay planes

LemonInTheDark committed:
Fixes some issues with excited group overlays

LemonInTheDark committed:
Adds a plane debugging ui

It lets you visualize how plane masters connect, edit their relays,
read details about each plane, and includes some in game documentation
for how plane masters work.

As a consequence of this, I've added a documentation variable on plane
masters. It forces devs to insert an explanation of what their plane
master does in text, partly so it can be displayed in text, and partly
just so it can be read.

Did some general plane cleanup, renamed backdrop() to show_to() because
backdrop() was NOT the right name, and the comment above it was HYPER
wrong.

LemonInTheDark committed:
Removes some unused code and cleans up some existing stuff

LemonInTheDark committed:
Adds offset support to mutable_appearance and emissives

It's actually only really needed when we manually set planes, which is
rare.
Need to do a pass on vis_contents and some other stuff, also uh, handle
on mob overlays. not sure what the heck I'ma do about that

LemonInTheDark committed:
Does a pass for underlays usage

Also cleans up some existing code, and ensures said existing code will
update on z move

LemonInTheDark committed:
Does a pass for vis_contents usage

I got most of them, but not the turf ones because I'm honestly not sure
how they'll work. I think I need to do some testing

Used vis_flags to apply inherit plane where appropriate

LemonInTheDark committed:
more todo

LemonInTheDark committed:
Haha johnathan I am dying (Resolves PM issues on map views)

Ok so like, there's this idea in byond, about submaps
Basically we can use a prefix before screen_loc values, and some byond
dmf fuckery to get a second map window to draw to.

We'd then setup some plane masters with appropriate screen locs, and use
vis_contents or just appearances to get an object with the appropriate
screen_loc to look like the thing we wanna mimic.

This is how we do camera consoles and such.

Basically, this system needs to account for offset planes. But as I was
working on it, it occured to me: shouldn't it also account for stuff
that modifies planes, basically shouldn't huds "know" what planes are
effecting them.

In persuit of this, I refactored how plane masters are handled, adding
plane groups, that represent planes rendered on some particular screen.

It in theory supports preloading plane masters, and applying them to
huds later, tho I have yet to find use for that.

In addition, life continues to be suffering, so I refactored all uses of
this pattern to use one consistent object, in the same rough way.

This makes the code a lot easier to reason about and read, and makes
making modifications to the pattern much easier

Note to self: If I'm gonna do this whole plane master group thing, I
might need a way to get existing "modifiacations" to plane masters

That or just nut up and make most things use the main group, not sure

Oh and I'm gonna make the debug window capable of editing these things,
for sniks

LemonInTheDark committed:
Adds support for viewing plane master groups to the ui

Disables it for now, because non main plane groups don't actually well,
work right now. They don't use relays, which is kinda the whole point of
the ui

Enabling it's a toggle tho, so hopefully it'll be easy to use in future

LemonInTheDark committed:
Subgroups are no longer used by effect sources, like controllers or say, lighting alpha

Too much risk of "dropping" a change, it's better this way
Oh and I documented a bunch of stuff, hopefully it'll help understanding

LemonInTheDark committed:
Cleans up map view logic a bit, makes gateways "work"

Ok so like, we can't do the old gateway effect cause vis planes aren't
being inherited
Anyway, instead, lets display that stuff in the ui, and bring back the
old effect for the portal being ready, give it a nice scanline effect

LemonInTheDark committed:
moth is bullied

LemonInTheDark committed:
Ripple doesn't do anything, ya yeeet

LemonInTheDark committed:
Adds checks for non existant spokesmen being passed into mut apper creation

Ensures limbs have spokesmen (most of the time)
Will need to speak with kapu about this

LemonInTheDark committed:
Adds logic to human overlays to allow for updating their planes "in place"

Ok so like, doing a full update would be too expensive, so instead we're
gonna create a list of all the overlays that either have a unique plane,
or have a child with a uniuqe plane

Then we walk up that list to remove overlays, and down it to add
Cause of the overlay flattening biz and all

Note: What I want to do in the end is store a copy of this list over
each iteration, rather then recalcing it each z change.

I can't atm tho, because when I try to I end up in fucky hell
Basically, overlays are weird, and there's something funky goin on.
I've seen behavior where appearances were pulled from the void, and
applied to the mob

Scares me

LemonInTheDark committed:
FUCK   YOU

LemonInTheDark committed:
Makes an attempt at including hud members into the plane thing

This'll need more work, but it's a start

LemonInTheDark committed:
Cleanliness, and makes the code compile

I'm gonna make huds not respect plane changes in future, but I want this
to be consistent so people don't fuck up by copying the wrong thing

LemonInTheDark committed:
Adds support for disabling offsets on some planes, also proper multiz darkness

Some planes really just shouldn't get stacked, mostly the hud related
ones

I've added a way to disable that in such a way that things attempting to
offset will just roll back to the main initial plane value
Useful.

I've also actually made the darkness plane matter, by adding the
see_blackness flag to all mobs
This lets me actually move around darkness, and shift it down or up
based on your z level.

LemonInTheDark committed:
cleans up some stuff, adds some todos, makes gas actually visible, pog

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
wahhh wahhh stop abusing macro- NO

LemonInTheDark committed:
fuck you

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Converts sight flags to a setter so I can react to them changing

This lets me get around blackness sticking around when it shouldn't
Oh also I've removed a lot of fluff from the blackness plane master
Doesn't seem to have needed it, and the blend mulitply broke slowly
changing alpha, which I'm not a fan of

LemonInTheDark committed:
Cleans up some sight sets

Also mandates see_blackness, cause you always need to be able to see it
for the pm to work.
Also removes a redundant sight flag thing from nightvision eyes

LemonInTheDark committed:
fuck off

LemonInTheDark committed:
wah wahhh

LemonInTheDark committed:
fixes c&d tests, adds some better errors, etc

LemonInTheDark committed:
fuck you again

LemonInTheDark committed:
Cleans up some todos, adds some new ones

LemonInTheDark committed:
cleans up some manual plane sets, I have a feeling this will be a pattern

LemonInTheDark committed:
ais can see lighting properly, adds a client.eye setter, adds some suffering to prospective code

LemonInTheDark committed:
Adds support for editing another mob's planes, for debug purposes

Makes it a lot easier to solve "hey what's going on" to be able to use
the nice ui to change stuff, rather then you know, doing it manually
with vv

LemonInTheDark committed:
Adds a setter and signal for see_invis

LemonInTheDark committed:
Adds a setter for see_in_dark, signals too, simiar reason to see_invisible

LemonInTheDark committed:
Adds support for limited mirroring of the view of one mob onto another

Ties it into the plane debugger ui, to make debugging player issues on
live easier

It's quite hacky, mostly because of how plane masters work, but there's
really not another way at the moment

Sorry

LemonInTheDark committed:
Fixes blur not applying to the floor, I am in fact a dumbass

LemonInTheDark committed:
Adds support for enabling/disabling mirroring in the ui

LemonInTheDark committed:
Removes all filtering from carbon updating logic

Overlay order effects rendering order, so if anything is filtered we run
the risk of dropping entries.
I might redo this at some point to drop inputs with no non floating
planes, but I feel like those will be rare so not right now

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
fine fine

LemonInTheDark committed:
bastards, fixes nonexisting parent calls

LemonInTheDark committed:
Adds support to multi cams, since they weren't working as expected

LemonInTheDark committed:
Ensures parallax works between z levels via mirage

Instead of positioning parallax objects on the lowest z level, we relay
them to all z levels
This way we don't need to move them, and they'll show up as expected

Note to self: add a way to edit relays and their blendmodes in future,
it'll be useful

Oh and I had to slightly rework how init relay adding is done, because
it didn't really support adding a relay with more data then just the
target plane
Not sure how I feel about this, I think it's alright though

LemonInTheDark committed:
haha get owned

LemonInTheDark committed:
Genericizes updating overlays, uses it to make statues work

LemonInTheDark committed:
Cleans up some appearance updates that weren't working great, added back a check for null locations being passed in

LemonInTheDark committed:
Moves emissive updating on init to hook into update_overlays

Overlays are not "like" removes like. if an overlay isn't added via
update_overlays, it can't be removed and refreshed automatically.

Emissive overlays were being added both in update_overlays, and in
initialize, which mean the one in initialize would never be removed

So every object that had one emissive blocker actually had two,
including the type that uses render targets.

sooooo lets just not do that one, and instead update appearance.

As a consequence, overrides of update_appearance and its subprocs need
to support being called at this stage in init, which many did not.

Oh and I've converted a lot of stuff that used the path -> datum in the
same var pattern, because it causes runtimes often with this method, and
is also really stupid

LemonInTheDark committed:
loosens the restrtictions on the mutable appearance check for planes, I wanna be nicer about this, needless mess for api users

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
whoops

LemonInTheDark committed:
wahhh wahh no var/ in proc args wahhh

LemonInTheDark committed:
fixes up all the errors caused by the update_appearance in atom initialize

LemonInTheDark committed:
lmao lol

LemonInTheDark committed:
ensures weather overlays are properly mutliz. It's possible that this will lead to improper double layering, but I really don't want to make 3 extra planes just for this effect

LemonInTheDark committed:
enables blocking render relay creation, uses this to ensure there's only ever one fov plane master

LemonInTheDark committed:
tmp

LemonInTheDark committed:
Lays a protective ward of appearance.overlays.Copy()

Without this, appearances will "stick" and be impossible to remove.
They will also occasionally "steal" or corrupt into the appearance of
something else on the map. I do not know why.

I'm in the process of building a testcase for this but I want to make
sure it works fully first

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
fixes emissive overlays with no source

LemonInTheDark committed:
nother CI thing

LemonInTheDark committed:
sorts plane masters by plane, removes redundant byondlightingplane define, it was removed 4 years ago when ai camera static plane was created

LemonInTheDark committed:
jesus christ who mutates behavior in fucking update_overlays. Should also stop treating one var as two values, come on man

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
resolves runtimes in ci, cleans a mistake

LemonInTheDark committed:
I am dumb

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Ensures FLOAT_PLANE is always properly managed by plane list mappings

As a consequence, resolves a bug with statue displaying
I added some extra code to make debugging it easier

Also solves a runtime in supermatter code

LemonInTheDark committed:
Fixes the podlauncher map not properly displaying

LemonInTheDark committed:
stacktraces in mirage destroy so I can see what's fucking the unit tests

LemonInTheDark committed:
WARNING WARNING TEMP

LemonInTheDark committed:
resolves random runtimes in unit tests

LemonInTheDark committed:
ahhhhhhhh

LemonInTheDark committed:
Cleans up SET_PLANE slightly, so most uses require less fluff

I'll document this soon

LemonInTheDark committed:
it is done

LemonInTheDark committed:
extra todo

LemonInTheDark committed:
Resolves a class of issue with hud images updating on loc move

Basically, we can handle images on basic huds updating loc generically
Since we have access to their loc on hud add, we can register off that,
and mange like that.

Also removes a redundant todo about ais and static through multiz, it's
an issue on master too

LemonInTheDark committed:
Resolves issues with updating some images on move. Nearly done with todos now

LemonInTheDark committed:
Removes redundant check. I don't wanna touch the rest of this until I'm ready for a fullmerge cause it'll be useful in debugging, so next is docs and then I'm gonna pr this thing

LemonInTheDark committed:
Adds documentation for EVERYTHING I know about our current rendering systems.

This includes my plane master changes, relays, appearances, and all the
little bits I think are useful to know/mention

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Mentions particles (thanks arm)

LemonInTheDark committed:
plane master controller snippet

LemonInTheDark commented:
Paging @Kylerace and @timothymtorres since they both seemed to understand what I was trying to do here

LemonInTheDark committed:
adds a link in contributing to the doc, fixes a compile error

LemonInTheDark committed:
Fixes gas overlays

LemonInTheDark committed:
updates static docs to reflect current reality

LemonInTheDark committed:
Fixes a hotlink

LemonInTheDark committed:
this fix it?

LemonInTheDark committed:
whoo[es

LemonInTheDark committed:
wahhh wahhh

optimumtact commented:
Probably the most important change since Intialize dropped, great work

optimumtact commented:
>The issue is how to get the actual atoms on the map to "land" on them properly.

NB: you don't actually explain how this is done, you hook creation and movement to place them into the appropriate planes based on z?

optimumtact commented:
also, how do openspace tiles know which plane to render as their contents?

MrStonedOne commented:
>AND it supports (in a very fragile and incomplete manner) viewing literally through someone else's eyes, including their plane masters. This is very useful, because it means you can debug "hey my X is yorked" issues yourself, on live.

This already exists on master.

The `observe` verb as a ghost will let you do this to any mob. View thru their hud, which controls their planemasters.

Did you preserve this?

Was your getters/setters work a duplication of the hud datum initialization system?

LemonInTheDark commented:
> NB: you don't actually explain how this is done, you hook creation and movement to place them into the appropriate planes based on z?

Whenever a plane variable is set (or on atom initialize/z move), I require some form of "spokesman" that can be used to get_turf the z level of the thing being set on. (outside niche cases)
We store a list of z level (as number) -> z level offset
Whenever a new z level is loaded, we check to see what level of offset it's in. the top level is 0, down 1 is 1, etc.

Then when we attempt to set plane, we index into that list on SSmapping with the z level, and get handed back the offset to use
Then it's just a matter of subtracting (offset * max plane delta), and we're done, things will land properly.

> also, how do openspace tiles know which plane to render as their contents?

Openspace tiles use vis_contents as before to tell byond to attempt to render the turfs below
The turfs below are already on lower turfs, so when they are rendered they'll be piped up through the set of render relays, onto their level's master plate, and then onto the transparent plane, so they'll be rendered visually "below"

> This already exists on master.
> 
> The `observe` verb as a ghost will let you do this to any mob. View thru their hud, which controls their planemasters.
> 
> Did you preserve this?
> 
> Was your getters/setters work a duplication of the hud datum initialization system?

the observe verb doesn't use the plane masters of the observed mob. I do.
This makes my code much more fragile, but lets me actually debug PM issues.
It ought? to all work, though I think what I did to sight might have yorked it slightly, will fix.

I don't think the hud datum actually tracks this stuff, so no?

LemonInTheDark committed:
Adds support for disabling plane masters

This doesn't mean alphaing them away, it means removing them and keeping
them out of our client's screen list.

This comes with a minor refactor of how relays are generated to make
some stuff easier.

With this, we are able to add excess PMs while not compromising client
performance anymore then 3xing the PM count does already.

I've applied this to the FOV mask plane, excited group overlays, ai
static, and pipecrawling, alongside the bottom level of openspace,
openspace vfx, and the transparent plane, as none of those will be used
at the bottom of a map.

This saves (3 * level depth) + 4 planes in the best case, or 13 out of
94 (icebox has 3 levels), or roughly 13% of our plane master count on icebox.
(it's 7 out of 34 in our best case on meta, or 20%. Not bad)

Not the worst thing in the world.

This does mean that these planes need to manually manage their
existence, but that's not that hard in these limited cases.

I've added support to it to the debug window too. That's about it for
now, I'd like to add support for disabling lower/higher level filters in
future, based on client prefs. Not sure how big an impact that would
actually make, we'll see.

LemonInTheDark commented:
I've added support for disabling temporary/limited (think ai static, pipecrawl, etc) plane masters. This hides them and their relays from our client screen, hopefully reducing client load.

This hides between 13% and 20% of our PM count, depending on the offset count.

I'm gonna look into adding a low specs pref and using it to disable things like filters on out of view planes. Not sure how large an impact this would actually make however.

LemonInTheDark committed:
runs prettier

LemonInTheDark commented:
Might not actually we don't have all that many filters and I'm not sure how big an impact it would actually make.
Would need to see people having issues before I dug that deep.

LemonInTheDark committed:
minor extra info about transform, minor bug fixes

LemonInTheDark committed:
null checks in plane group check

LemonInTheDark committed:
Attempts to optimize plane setting for the single z, init case

Time-Green commented:
Multi-z levels look a bit too dark
![image](https://user-images.githubusercontent.com/7501474/184090840-1c70e8ac-288e-45a5-9fcb-839318bafdf0.png)

![image](https://user-images.githubusercontent.com/7501474/184091120-d5d6c803-7b57-47d3-a486-5c035b78fd0f.png)

LemonInTheDark commented:
Yeet
it's a consequence of how we signify depth, that little dark bit we overlay on multiz
I'm working on a different multiz implementation so we can drop that, since it only worked before because lighting was yorked

LemonInTheDark committed:
Some minor bugfixes, wtf floyd

ShizCalev commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -130,23 +130,21 @@
 
 /obj/item/paper_bin/examine(mob/user)
 	. = ..()
+	var/total_paper = LAZYLEN(papers)
```
this doesn't seem relevant to this PR?

LemonInTheDark replied:
one would think
I setup code that calls update_overlays on init, for reasons, and since paperbins used the var this replaced as both a cache of the paper count AND a way to tell themselves on init how much paper to spawn, it broke
Fuck I forgot to mention that in the pr. I will.

https://github.com/tgstation/tgstation/pull/67902#issuecomment-1191952086

LemonInTheDark replied:
https://github.com/LemonInTheDark/tgstation/commit/644813af784dc9b5a900e2d4d3f6dfad176fc54f

ShizCalev commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -14,7 +14,7 @@
 	pressure_resistance = 8
 	var/papertype = /obj/item/paper
-	var/total_paper = 30
+	var/initial_paper = 30
```
.

ShizCalev commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -33,8 +33,8 @@
-	for(var/i in 1 to total_paper)
-		papers.Add(generate_paper())
+	for(var/i in 1 to initial_paper)
+		papers += generate_paper()
```
.

ShizCalev commented on code/modules/paperwork/paperbin.dm:
```diff
@@ -130,23 +130,21 @@
 
 /obj/item/paper_bin/update_icon_state()
-	if(total_paper < 1)
+	if(!LAZYLEN(papers))
```
.

ShizCalev commented on code/modules/paperwork/paperplane.dm:
```diff
@@ -57,7 +57,7 @@
 /obj/item/paperplane/update_overlays()
 	. = ..()
-	for(var/stamp in internalPaper.stamp_cache)
+	for(var/stamp in internalPaper?.stamp_cache)
```
.

ShizCalev commented on _maps/RandomRuins/SpaceRuins/onehalf.dmm:
```diff
@@ -680,7 +680,7 @@
 /obj/structure/table/reinforced,
 /obj/item/paper_bin{
-	total_paper = 12
+	initial_paper = 12
```
.

ShizCalev commented on _maps/shuttles/whiteship_donut.dmm:
```diff
@@ -399,7 +399,7 @@
 /obj/effect/decal/cleanable/dirt/dust,
 /obj/item/paper_bin{
-	total_paper = 20
+	initial_paper = 20
```
.

ShizCalev commented on code/game/objects/items/devices/portable_chem_mixer.dm:
```diff
@@ -67,7 +67,7 @@
 
 /obj/item/storage/portable_chem_mixer/update_icon_state()
-	if(!atom_storage.locked)
+	if(!atom_storage?.locked)
```
?

LemonInTheDark replied:
relevant to above, this can go null

ShizCalev commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -111,8 +113,8 @@
-	if(ispath(trim))
-		SSid_access.apply_trim_to_card(src, trim)
+	if(ispath(trim_path) && !trim)
+		SSid_access.apply_trim_to_card(src, trim_path)
```
what does this have to do with multiz

LemonInTheDark replied:
https://github.com/LemonInTheDark/tgstation/commit/644813af784dc9b5a900e2d4d3f6dfad176fc54f

LemonInTheDark replied:
Caused runtimes in init, and is also just annoying

ShizCalev commented on code/game/objects/structures/headpike.dm:
```diff
@@ -40,6 +40,9 @@
+	if(!victim)
+		name = initial(name)
+		return ..()
 	name = "[victim.real_name] on a [spear]"
```
name = victim ? "[victim.real_name] on a [spear]" : initial(name)

LemonInTheDark replied:
idk, if you're sure. I'm not a huge fan of ternaries but maybe it's fine here

ShizCalev commented on code/game/machinery/lightswitch.dm:
```diff
@@ -40,19 +40,19 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/light_switch, 26)
 
 /obj/machinery/light_switch/update_icon_state()
-	set_light(area.lightswitch ? 0 : light_on_range)
+	set_light(area?.lightswitch ? 0 : light_on_range)
```
@optimumtact iirc you yelled at me two years ago for trying to add this exact check lol

I assume this is encountered during initialize / destroy?

optimumtact replied:
u little snitch lol

it seems very unlikely area would ever be null in normal situations, so I still don't agree with the nullability check

LemonInTheDark replied:
I thought about it a bit, I think it'd be better to check for the initialized flag in gases like this 
since we don't want these cases running until we're out of initialize, we should just make that explicit

LemonInTheDark replied:
god fucking damn it I can't. SHIT

LemonInTheDark replied:
This sort of thing is really annoying because it's "this breaks if called before init is **mostly** done"
we need to be able to call it IN init, but we can't run it before area has been set, because if we do we get runtimes.
I can't even use the flag because the call I'm handing happens AFTER the flag is set

help

LemonInTheDark replied:
Resolved, I didn't actually need to call update_appearance

ShizCalev commented on code/game/objects/items/tools/weldingtool.dm:
```diff
@@ -190,7 +190,7 @@
 /// Returns the amount of fuel in the welder
 /obj/item/weldingtool/proc/get_fuel()
-	return reagents.get_reagent_amount(/datum/reagent/fuel)
+	return reagents?.get_reagent_amount(/datum/reagent/fuel)
```
Honestly I'd prefer explicitly checking if !initialized || qdeleting and returning on that since this is more likely to end up hiding different problems

ShizCalev commented on code/game/objects/items/cards_ids.dm:
```diff
@@ -111,8 +113,8 @@
-	if(ispath(trim))
-		SSid_access.apply_trim_to_card(src, trim)
+	if(ispath(trim_path) && !trim)
+		SSid_access.apply_trim_to_card(src, trim_path)
```
I'm guessing this runtime that I've been seeing on sybil is what this fixes? Been eyeing it for a few days

```
[2022-08-12 09:53:52.419] runtime error: Cannot read /datum/id_trim/job/prisoner (/datum/id_trim/job/prisoner).sechud_icon_state
 - proc name: get trim sechud icon state (/obj/item/card/id/proc/get_trim_sechud_icon_state)
 -   source file: cards_ids.dm,790
 -   usr: Helvetica8563 (/mob/dead/new_player)
 -   src: the prisoner ID card (/obj/item/card/id/advanced/prisoner)
 -   usr.loc: the plating (8,174,1) (/turf/open/floor/plating)
 -   src.loc: Haliee Mason (/mob/living/carbon/human/dummy)
 -   call stack:
 - the prisoner ID card (/obj/item/card/id/advanced/prisoner): get trim sechud icon state()
 - the prisoner ID card (/obj/item/card/id/advanced/prisoner): get trim sechud icon state()
 - the prisoner ID card (/obj/item/card/id/advanced/prisoner): get sechud job icon state()
 - Haliee Mason (/mob/living/carbon/human/dummy): sec hud set ID()
 - Haliee Mason (/mob/living/carbon/human/dummy): equip to slot(the prisoner ID card (/obj/item/card/id/advanced/prisoner), 256, 1, 0)
 - Haliee Mason (/mob/living/carbon/human/dummy): equip to slot if possible(the prisoner ID card (/obj/item/card/id/advanced/prisoner), 256, 1, 1, 0, 1, 1)
 - Haliee Mason (/mob/living/carbon/human/dummy): equip to slot or del(the prisoner ID card (/obj/item/card/id/advanced/prisoner), 256, 1)
 - Prisoner (/datum/outfit/job/prisoner): equip(Haliee Mason (/mob/living/carbon/human/dummy), 1)
 - Haliee Mason (/mob/living/carbon/human/dummy): equipOutfit(/datum/outfit/job/prisoner (/datum/outfit/job/prisoner), 1)
 - Haliee Mason (/mob/living/carbon/human/dummy): dress up as job(/datum/job/prisoner (/datum/job/prisoner), 1)
 - ...
 - /datum/preferences (/datum/preferences): create character preview view(Helvetica8563 (/mob/dead/new_player))
 - /datum/preferences (/datum/preferences): ui data(Helvetica8563 (/mob/dead/new_player))
 - /datum/tgui (/datum/tgui): get payload(null, 1, 1)
 - /datum/tgui (/datum/tgui): open()
 - /datum/preferences (/datum/preferences): ui interact(Helvetica8563 (/mob/dead/new_player), /datum/tgui (/datum/tgui))
 -  (/atom/movable/screen/lobby/button/character_setup): Click(null, "mapwindow.map", "icon-x=61;icon-y=21;left=1;but...")
 -  (/atom/movable/screen/lobby/button/character_setup):  Click(null, "mapwindow.map", "icon-x=61;icon-y=21;left=1;but...")
 - /datum/callback/verb_callback (/datum/callback/verb_callback): Invoke()
 - /datum/callback/verb_callback (/datum/callback/verb_callback): InvokeAsync()
 - 
```

LemonInTheDark replied:
Likely yes. trim stored both the path, and the object created from the path. the bug there could just be trying to get icon state before the trim is new'd, but I'm not sure when that ordinarily happens

LemonInTheDark committed:
Sets up turf transparency to implement a modification of lummy's pillar idea. Does not work because getting pillars in view is VERY hard

LemonInTheDark committed:
removes is_openspace arg, changes some planes

LemonInTheDark committed:
ensures drones can no longer passthrough doors

LemonInTheDark committed:
adds support for detached plane masters to the debugger, also makes masters more distinct as the start/end of lines

LemonInTheDark committed:
reworks how transparency works because pillars were NOT cooperating with me

LemonInTheDark committed:
Reorders plane masters a bit, adds some new render plates

LemonInTheDark committed:
whoops forgot a spot

LemonInTheDark committed:
adds a translate pass when you move up/down z levels that zooms in and out plane masters

LemonInTheDark committed:
datumizes z pillars, cleans up z holder logic a bit

LemonInTheDark committed:
Cleans up some code, makes pillars actually transparent, adds some todos:

LemonInTheDark committed:
Adds multiz lighting

Basically, if a tile is potentially multiz, we add any turfs above/below
transparent tiles to our turfs list, and edit the corner's intensity
based off z as well as just normal x/y dist.

I've tried to set this up to make it as minimal as possible in terms of
impact on not multiz. there's slight overhead, but it's not THAT bad

LemonInTheDark committed:
Sets up the clickcatcher plane to travel up/down with you

I can't "remove" clickcatcher plane masters once they become redundant,
but I can make them have no cost but memory so eh

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
whoops

LemonInTheDark committed:
hhhh

LemonInTheDark committed:
fixes a warning

LemonInTheDark committed:
fixes an infinite loop, ensures some things are setup properly

LemonInTheDark committed:
cleans up lighting code to use z diffs rather then a source to source thing. This is roughly as fast, and correct. Also I changed the equation a bit, since cube roots are too aggressive here

LemonInTheDark committed:
Removes debug code, adds a comment explaining the logic a bit

LemonInTheDark committed:
todos and given up on todos

LemonInTheDark committed:
Renames accepts_input to better match its real use

LemonInTheDark committed:
cleans up a unit test runtime, adds another todo

LemonInTheDark committed:
you saw nothing

LemonInTheDark committed:
Fixes up transparency to actually clear out

Also ensures our abstract things don't move off shuttles, and adds some
todos

LemonInTheDark committed:
todone

LemonInTheDark committed:
Replaces the atmos plane master with a high_game_plane PM

This contains atmos overlays, mining overlays, and fov stuff.

Basically anything we want to display and not be touchable, but
that has to support different z layers.

This is done because I do not want to have to manage screen planes, and
would prefer to just assert that there won't be one that's on a non
absolute plane.

LemonInTheDark committed:
clarification, fixes clickcatcher

LemonInTheDark committed:
fixes ghosts spawning on the wrong plane

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
whoops

LemonInTheDark committed:
Fixes a 'before init' error in ants

timothymtorres commented:
Possible issues this might solve:

- #64385
- #64098
- #62442
- #61968
- #59610
- #56390
- #53783
- #53531
- #50444
- #49890
- #45669
- #40542
- #39799
- #67336
- #37303
- #53581
- #68461
- #68154

msgerbs commented:
Could this also potentially fix this pesky bastard?

* [ Welding masks intermittently not blocking vision #52972](https://github.com/tgstation/tgstation/issues/52972)

LemonInTheDark commented:
I'm sorry but I'm gonna really disappoint you, I don't think I've hit most of these

* [#64279 layered borg eyelights above mobs and cover-overlays #64385](https://github.com/tgstation/tgstation/issues/64385) Nope, likely a consequence of eyelights using the above lighting plane rather them emissives

* [Emissives do not change when the wearer lays down #64098](https://github.com/tgstation/tgstation/issues/64098) See below

* [emissive blockers aren't applied properly for both body parts and worn overlays #62442](https://github.com/tgstation/tgstation/issues/62442) Also nope, this is likely just the reset_transform fuckin us up

* [Overlays from BCI going under tiles(?) #61968](https://github.com/tgstation/tgstation/issues/61968) honestly no fucking idea what's wrong here, outside of those things not being on the hud plane. what even ARE they can't find em in the code

* [Cryo tubes don't always seem to get rid of the occupant (overlay?) when the occupant is ejected #59610](https://github.com/tgstation/tgstation/issues/59610) I could swear we fixed this already, I guess not

* [Skyburger flattened into singularity #56390](https://github.com/tgstation/tgstation/issues/56390) this isn't really a bug, tho someone do please make really tall burgers shake and have an approaching 100% chance to fall over as they get taller

* [Mesons allow seeing objects, not just turfs, in the z-level below you through walls #53783](https://github.com/tgstation/tgstation/issues/53783) This would be potentially fixable now that we have control over the blackness plane, but I've not done so yet

* [Radiation storm overlay interferes with runechat #53531](https://github.com/tgstation/tgstation/issues/53531) I may have fixed this with my weather effects pr, honestly not sure how it was happening in the first place. doesn't happen now tho.

* [When Ventcrawling Everything Is Dark. #50444](https://github.com/tgstation/tgstation/issues/50444) I believe this is because slimes have a low see_invisible value, which masks away the fullbright and fulldark images. not totally sure, but that's my assumption. not fixed here

* [Mob overlays blur when mobs are resized #49890](https://github.com/tgstation/tgstation/issues/49890) Local man doesn't place image in issue report. Not sure what this means, as scaling a human works fine, and they are all overlay. needs more detail, not fixed by me

* [Some screens render over Nar'Sie #45669](https://github.com/tgstation/tgstation/issues/45669) I think the fix for this would be making narsie block emissives? Not sure

* [Mining scanners with mesons during ash storms leave "empty" tiles in place of unexposed ores instead of "rock" tiles #40542](https://github.com/tgstation/tgstation/issues/40542) ... WHAT? I cannot replicate this.

* [You can see items through walls with engineering meson scanner. #39799](https://github.com/tgstation/tgstation/issues/39799) We use an image for this, with a location of the target turf. This might be fixed just by using an overlay instead. Not sure

* [The light / emissive effect of beams are not aligned properly #67336](https://github.com/tgstation/tgstation/issues/67336) Nope. Same bug as before, emisssives refuse to rotate. Why? I have no fucking clue

* [Images on guns don't rotate when thrown #37303](https://github.com/tgstation/tgstation/issues/37303) This is interesting, they do rotate, but not in the same way as the gun, they refuse to turn. Not fixed here

* [Wall Smoothing issues on MultiZ #53581](https://github.com/tgstation/tgstation/issues/53581) Fixed yes. I want to redo how this works actually, the transparent element should be used here but it's been hacked to work for glass walls, so I'll need to fix that first. Maybe later, I wanna be able to see under top level shuttles damn it. 

* [Glass floors are made of pure suffering #68461](https://github.com/tgstation/tgstation/issues/68461) Fixed. Honestly not sure why? maybe it was the vis_contents rules of the gas overlay. But fixed.

* [alien resin sometimes doesn't change colors during timestop? #68154](https://github.com/tgstation/tgstation/issues/68154) Nope. I have no idea what would cause this, makes me think something weird with atom color's goin on

LemonInTheDark committed:
ensures the tram works, and removes the need to mass update_appearance in init

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
completes todo

LemonInTheDark committed:
Revert "jesus christ who mutates behavior in fucking update_overlays. Should also stop treating one var as two values, come on man"

This reverts commit 1c28e60e327dc2ed9b9231033c0fe5652d907563.

LemonInTheDark committed:
pain

LemonInTheDark committed:
removes now redundant checks

LemonInTheDark committed:
Revert "Moves emissive updating on init to hook into update_overlays"

This reverts commit 644813af784dc9b5a900e2d4d3f6dfad176fc54f.

LemonInTheDark committed:
makes it all compile. Init is a bit high, check that closer. also uh, space fullbright is WEIRD, so do some touching with that too

optimumtact commented on code/modules/industrial_lift/industrial_lift.dm:
```diff
@@ -17,6 +17,8 @@ GLOBAL_LIST_EMPTY(lifts)
 	obj_flags = CAN_BE_HIT | BLOCK_Z_OUT_DOWN
 	appearance_flags = PIXEL_SCALE|KEEP_TOGETHER //no TILE_BOUND since we're potentially multitile
+	// If we don't do this, we'll build our overlays early, and fuck up how we're rendered
+	blocks_emissive = 0
```
NONE

LemonInTheDark replied:
aight

LemonInTheDark commented:
I still have some stuff I need to work out with this, outside the obvious. space looks bad right now, and I'm worried about init times
Also uh, full transparency, I added parallaxing to multiz, and multiz lighting. things looked like shit without it, and it made me sad

if that's an issue to bundle with this I can atomize it out or whatever

LemonInTheDark committed:
sets up fullscreen following

LemonInTheDark committed:
implements turf to turf area lighting, but only for turfs NOT on the prime layer

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Fixes some stuff I missed

LemonInTheDark commented:
I think init is fine with this. We gained maybe 0.5 seconds on space init? but nothing that bad, that much was slightly expected

LemonInTheDark committed:
NONE

LemonInTheDark commented:
TM'd on sybil, just to be sure

LemonInTheDark committed:
fixes up part of ci

LemonInTheDark committed:
will this help?

LemonInTheDark committed:
please don't update appearance if you're qdeleted brother

LemonInTheDark committed:
Reworks how we track area z levels

This makes space init slightly slower, but speeds up turf init. We
needed to slow space init anyway, so what can you do.

LemonInTheDark committed:
lowers init time by not being stupid as hell

LemonInTheDark committed:
another todo :(

ShizCalev commented:
```
[14:40:05] Runtime in stack_trace.dm, line 3: No plane offset passed in as context for a non floating mutable appearance, ya done fucked up
proc name: stack trace (/proc/stack_trace)
usr: BNQORD/(Joe Bicknell)
usr.loc: (Fore Primary Hallway (100,154,4))
src: null
call stack:
stack trace("No plane offset passed in as c...")
mutable appearance('icons/obj/clothing/suits/ether...', "eth_raincoat_glow", -1, null, 14, 255, 106, null)
emissive appearance('icons/obj/clothing/suits/ether...', "eth_raincoat_glow", null, -1, 255, 0, null)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): update overlays(16777215)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): update icon(16777215)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): update appearance(16777215)
the ethereal raincoat (/obj/item/clothing/suit/hooded/ethereal_raincoat): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, 1)
Joe Bicknell (/mob/living/carbon/human): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, 1)
Joe Bicknell (/mob/living/carbon/human): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, null)
Joe Bicknell (/mob/living/carbon/human): on changed z level(the plating (99,154,3) (/turf/open/floor/plating), the floor (100,154,4) (/turf/open/floor/iron), 0, null)
...
the plating (99,154,3) (/turf/open/floor/plating): SendSignal("atom_exit", /list (/list))
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 4, 0)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 4, null)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 4)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 6, null)
Joe Bicknell (/mob/living/carbon/human): Move(the wall (100,154,3) (/turf/closed/wall), 6)
Joe Bicknell (/mob/living/carbon/human): keyLoop(BNQORD (/client))
Input (/datum/controller/subsystem/verb_manager/input): ignite(0)
Master (/datum/controller/master): Loop(2)
Master (/datum/controller/master): StartProcessing(0)
```

scriptis commented:
Camera consoles appear to be extremely broken, uncertain if it's related to this.

LemonInTheDark commented:
would totally be related, thank you. will look into it

ShizCalev commented:
```
[18:28:04] Runtime in stack_trace.dm, line 3: No plane offset passed in as context for a non floating mutable appearance, ya done fucked up
proc name: stack trace (/proc/stack_trace)
usr: Normmacdonaldfan/(Silent Retard)
usr.loc: (Starboard Primary Hallway (142,110,2))
src: null
call stack:
stack trace("No plane offset passed in as c...")
mutable appearance('icons/effects/effects.dmi', "thought_bubble", -1, null, 5, 255, 64, null)
Silent Retard (/mob/living/carbon/human): create point bubble(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): point at(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): pointed(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): pointed(the pen (/obj/item/pen))
/datum/callback/verb_callback (/datum/callback/verb_callback): InvokeAsync()
Silent Retard (/mob/living/carbon/human): Point To(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): Point To(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): ShiftMiddleClickOn(the pen (/obj/item/pen))
Silent Retard (/mob/living/carbon/human): ClickOn(the pen (/obj/item/pen), "icon-x=14;icon-y=14;middle=1;s...")
the pen (/obj/item/pen): Click(null, "mapwindow.map", "icon-x=14;icon-y=14;middle=1;s...")
the left hand (/atom/movable/screen/inventory/hand): Click(null, "mapwindow.map", "icon-x=14;icon-y=14;middle=1;s...")
Normmacdonaldfan (/client): Click(the left hand (/atom/movable/screen/inventory/hand), null, "mapwindow.map", "icon-x=14;icon-y=14;middle=1;s...")
```

LemonInTheDark commented:
> Camera consoles (and other embedded views, such as pod launch previews) appear to be extremely broken, uncertain if it's related to this.

I can fix this, but it either means breaking multiz parallax, or forcing a client version update so I can be ensured some byond features work

LemonInTheDark committed:
spokesman's thought bubbles, cleans them up, makes the point place mouse transparent

LemonInTheDark committed:
Fixes mapviews for just modern clients. This might require a fversion upgrade. This is required because of how I'm shifting lighting to make plane master transforms for multiz work

Time-Green commented:
> > Camera consoles (and other embedded views, such as pod launch previews) appear to be extremely broken, uncertain if it's related to this.
> 
> I can fix this, but it either means breaking multiz parallax, or forcing a client version update so I can be ensured some byond features work

Force them to update, I wont be able to live without the multiz parallax

dragomagol commented:
It sounds like this is causing radiation storms/ash storms to be invisible again

LemonInTheDark committed:
Fixes weather visuals, lowers z containing list size

LemonInTheDark commented:
Handled. this still has some bugs tho, with multiz and also camera consoles
workin on em.

Baskett55 commented:
a rod spawned in lower levels of tram and it was invisible, like it was underneath the space tiles, until it reach the station and it looked normal.

LemonInTheDark commented:
I found another fucking byond bug
this is now number 3 
I hate it here

LemonInTheDark committed:
temporary fix for space, this NEEDS to be optimized in more detail. I'm sorry moth

ShizCalev commented on code/game/alternate_appearance.dm:
```diff
@@ -65,14 +65,15 @@ GLOBAL_LIST_EMPTY(active_alternate_appearances)
 	transfer_overlays = options & AA_MATCH_TARGET_OVERLAYS
 	image = I
 	target = I.loc
+	RegisterSignal(target, COMSIG_MOVABLE_Z_CHANGED, .proc/target_z_changed)
```
[21:42:15] Runtime in stack_trace.dm, line 3: movable_ztransit overridden. Use override = TRUE to suppress this warning

msgerbs commented:
I believe #69968 was probably caused by this as well, have you seen it?

LemonInTheDark commented:
Yeah it was. It's not got a fix PUSHED yet but I've got it

LemonInTheDark commented:
WAKE ME UP INSIDE https://www.byond.com/forum/post/2821969

LemonInTheDark committed:
Disables multiz parallax when you can see turfs but not objects

BYOND reasons. It don't work with how I do edges, and the workaround
(images) is broken for similar but unrelated reasons. Pain

LemonInTheDark committed:
Changes camera static to not be dumb

Camera static now supports multiz
Chunks now cover the full z stack, and cameras will show turfs
below/above transparent turfs
Oh also I cleaned up chunk code

LemonInTheDark committed:
Cleans up how we manage transforms for plane masters

Should be on hud 4head, this is what god invented signals for

LemonInTheDark committed:
save meeee

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
Removes rendundant blend_mode sets alongside changing what sort of
client color handing is default

It's nicer this way

LemonInTheDark committed:
whoops

LemonInTheDark committed:
KILL MEEEE

LemonInTheDark committed:
Fucking camera code

LemonInTheDark committed:
cleans up some todos:

LemonInTheDark committed:
Removes parallax from popup windows, it's hard to manage and I really need to finish this

LemonInTheDark committed:
Optimizes turfs and space turf offsetting

LemonInTheDark committed:
Fixes render relay hard dels

LemonInTheDark committed:
Fixes up some hard deletes with render relays, plane masters and popup planes

LemonInTheDark committed:
cleans up some todos, unrolls a define

LemonInTheDark committed:
fixes two runtimes, one of which was my fault

LemonInTheDark committed:
whoops

LemonInTheDark committed:
Moves over debug procs, better documents my hellspawn

LemonInTheDark committed:
documents things, cleans up some code, renames some files

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

Time-Green commented:
> I'm close to done with this.
> Need to update the pr body and push a few last things, and then we are golden (I hope)
> Oh also test init times so I can make peace with my moth overlord

The future of spaceman is upon us!!!

LemonInTheDark commented:
There's so much work left to be done brother

LemonInTheDark committed:
Removes zs we have, don't need it and it just ate cpu time. Optimizes away maybe 0.4 seconds of turf init by being not stupid TM

LemonInTheDark committed:
Whoops

LemonInTheDark committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into multiz-hell

LemonInTheDark committed:
can't believe I just left this

LemonInTheDark committed:
makes multiz lighting actually work

Time-Green commented:
Oh god oh fuck jesus christ it’s happening

LemonInTheDark commented:
fuck I forgot to update the desc
ah well
![plane_cube](https://user-images.githubusercontent.com/58055496/192466699-1c461e3e-1bfc-4c36-bea2-b91b2923d9a6.png)
